pkg/streamingpromql/operators/aggregations/limitklimitratio/operator.go

```diff
@@ -0,0 +1,432 @@
+
+		counter, err := newStepCounter(o.TimeRange.StepCount, o.MemoryConsumptionTracker)
+		if err != nil {
+			return nil, err
+		}
+
+		return &groupLimiterK{
+			stepCounter: counter,
+			stepArg:     stepArg,
+			stepsFilled: zeros,
+		}, nil
+	}
+
+	return zeros < o.TimeRange.StepCount, nil
+}
+
+func (o *Operator) SeriesMetadata(ctx context.Context, matchers types.Matchers) ([]types.SeriesMetadata, error) {
+
+	// parse and validate the given input argument for this aggregate (ie k or ratio)
+	// also initialises the factory method for creating new groupLimiters
+	canReturnAnyResults, err := o.initParam(ctx)
+	if err != nil {
+		return nil, err
+	} else if !canReturnAnyResults {
+		// all the k/ratio values are 0 so we will have nothing to return
+		return nil, nil
+	}
+
+	innerSeries, err := o.Inner.SeriesMetadata(ctx, matchers)
+	if err != nil {
+		return nil, err
+	}
+
+	groupLabelsBytesFunc := aggregations.GroupLabelsBytesFunc(o.Grouping, o.Without)
+	groups := map[string]groupLimiter{}
+
+	o.seriesToGroups = make([]groupLimiter, len(innerSeries))
+
+	// Go through each series and determine which group the series will belong to.
+	// A limitGroup is created for each unique group, which will be used to assist in tracking the series values across the whole group at each step.
+	// Note that we can not discard any series at this point, even if we observe more series then the k limit, as we do not yet know which series have data.
+	// A series with no samples will not contribute to increasing k.
+	for idx, series := range innerSeries {
+		groupLabelsString := groupLabelsBytesFunc(series.Labels)
+		g, groupExists := groups[string(groupLabelsString)] // Important: do not extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if !groupExists {
+			g, err = o.groupLimiterFactory()
+			if err != nil {
+				return nil, err
+			}
+			groups[string(groupLabelsString)] = g
+		}
+
+		g.increaseSeriesCount()
+		o.seriesToGroups[idx] = g
+	}
+
+	// Perform a second iteration of all the series to initialize each series into its group.
+	// This is done as a second step since now we know how many series are in each group.
+	// This is only done for limit_ratio, as this step allows the caching of the series label hash
+	for idx, g := range o.seriesToGroups {
+		g.initSeries(idx, innerSeries[idx])
+	}
+
+	return innerSeries, nil
+}
+
+func (o *Operator) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+
+	if o.nextSeriesIndex >= len(o.seriesToGroups) {
+		return types.InstantVectorSeriesData{}, types.EOS
+	}
+
+	defer func() {
+		// Note this will also close() the group if it's series ref count has reached 0
+		o.seriesToGroups[o.nextSeriesIndex].releaseSeries(o.nextSeriesIndex)
+		// clear these references to allow the GC to clean up sooner
+		o.seriesToGroups[o.nextSeriesIndex] = nil
+		o.nextSeriesIndex++
+	}()
+
+	g := o.seriesToGroups[o.nextSeriesIndex]
```

Comment:
[nit] Might be clearer to move this above the `defer` so that the `defer` can use `g` too. Then these lines could move out of the `defer`: 

```go
		o.seriesToGroups[o.nextSeriesIndex] = nil
		o.nextSeriesIndex++
```